File: test/unit/parser/function.rb
#
# Copyright 2011 by Salman Ahmad (salman@salmanahmad.com).
# All rights reserved.
#
# Permission is granted for use, copying, modification, distribution,
# and distribution of modified versions of this work as long as the
# above copyright notice is included.
#

require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', 'test_helper.rb'))

class ParserTests::FunctionTest < Test::Unit::TestCase
  
  def setup
    @parser = Dog::Parser.new
    @parser.parser.root = :function_definition
  end
  
  def test_simple
    program = <<-EOD
DEFINE function DO 
  i
END
EOD
    
    program.strip!
    @parser.parse(program)
    
  end
  
  
  
  def test_nested
    program = <<-EOD
DEFINE foo DO 
  DEFINE bar DO
    DEFINE bubble DO
  
    END
  END
  
  DEFINE baz DO
  
  END
END
EOD
    
    program.strip!
    @parser.parse(program)
    
  end
  
  
  def test_on_and_using
    
    
    program = <<-EOD
DEFINE function ON input DO 




END
EOD
    
    program.strip!
    @parser.parse(program)
    
    program = <<-EOD
DEFINE function ON input USING config DO 




END
EOD

    program.strip!
    @parser.parse(program)
    
  end
  
  
  
  
  
  
  
  
end

File: src/main/java/dog/lang/compiler/Function.java
/*
 *
 *  Copyright 2012 by Salman Ahmad (salman@salmanahmad.com).
 *  All rights reserved.
 *
 *  Permission is granted for use, copying, modification, distribution,
 *  and distribution of modified versions of this work as long as the
 *  above copyright notice is included.
 *
 */

package dog.lang.compiler;

import dog.lang.nodes.Node;
import dog.lang.instructions.Instruction;
import dog.lang.Resolver;

import org.objectweb.asm.*;

public class Function extends Symbol implements Opcodes {
	public Function(String name, Node node, Compiler compiler) {
		super(name, node, compiler);
	}

	public String toDogBytecodeString() {
		String output = "";
		output += String.format("; function: %s\n", name);
		output += String.format("; variables: %d stack: %d\n", this.variableGenerator.variableCount(), registerGenerator.registerCount());

		for(int i = 0; i < instructions.size(); i++) {
			output += String.format("%04d %s\n", i, instructions.get(i).toString());
		}

		output += "; end function\n\n\n";

		return output;
	}

	public void compile() {
		compileContinuable("dog/lang/Function");
	}
}
